src/vm/method.cpp

```diff
@@ -4959,84 +4959,80 @@ LPVOID NDirectMethodDesc::FindEntryPoint(HINSTANCE hMod) const
+    int dstbufsize = (int)(sizeof(char) * (strlen(funcName) + 1 + 20)); // +1 for the null terminator
                                                                          // +20 for various decorations
 
     // Allocate a single character before the start of this string to enable quickly
     // prepending a '_' character if we look for a stdcall entrypoint later on.
-    LPSTR szAnsiEntrypointName = ((LPSTR)_alloca(dstbufsize + 1)) + 1;
+    LPSTR szProbedEntrypointName = ((LPSTR)_alloca(dstbufsize + 1)) + 1;
 
     // Copy the name so we can mangle it.
-    strcpy_s(szAnsiEntrypointName,dstbufsize,GetEntrypointName());
-    DWORD nbytes = (DWORD)(strlen(GetEntrypointName()) + 1);
-    szAnsiEntrypointName[nbytes] = '\0'; // Add an extra '\0'.
-
+    strcpy_s(szProbedEntrypointName, dstbufsize, funcName);
+    DWORD probedFuncNameLength = (DWORD)(strlen(funcName) + 1);
+    szProbedEntrypointName[probedFuncNameLength] = '\0'; // Add an extra '\0'.
 
-    // If the program wants the ANSI api or if Unicode APIs are unavailable.
-    if (IsNativeAnsi())
+    if(!IsNativeNoMangled())
     {
-        szAnsiEntrypointName[nbytes-1] = 'A';
-        pFunc = GetProcAddress(hMod, funcName = szAnsiEntrypointName);
-    }
-    else
-    {
-        szAnsiEntrypointName[nbytes-1] = 'W';
-        pFuncW = GetProcAddress(hMod, szAnsiEntrypointName);
+        szProbedEntrypointName[probedFuncNameLength - 1] = IsNativeAnsi() ? 'A' : 'W';
 
-        // This overrides the unmangled API. See the comment above.
-        if (pFuncW != NULL)
+        FARPROC pProbedFunc = GetProcAddress(hMod, szProbedEntrypointName);
+
+        if (pProbedFunc != NULL)
         {
-            pFunc = pFuncW;
-            funcName = szAnsiEntrypointName;
+            pFunc = pProbedFunc;
         }
-    }
 
-    if (!pFunc)
-    {
+        probedFuncNameLength++;
+    }
 
 #if defined(_TARGET_X86_)
-        /* try mangled names only for __stdcalls */
-        if (!pFunc && IsStdCall())
+    /* try mangled names only for __stdcalls */
+    if (!pFunc && IsStdCall())
+    {
+        UINT16 numParamBytesMangle = GetStackArgumentSize();
+            
+        if (IsStdCallWithRetBuf())
         {
-            UINT16 numParamBytesMangle = GetStackArgumentSize();
-                
-            if (IsStdCallWithRetBuf())
-            {
-                _ASSERTE(numParamBytesMangle >= sizeof(LPVOID));
-                numParamBytesMangle -= (UINT16)sizeof(LPVOID);
-            }
-
-            szAnsiEntrypointName[-1] = '_';
-            sprintf_s(szAnsiEntrypointName + nbytes - 1, dstbufsize - (nbytes - 1), "@%ld", (ULONG)numParamBytesMangle);
-            pFunc = GetProcAddress(hMod, funcName = szAnsiEntrypointName - 1);
+            _ASSERTE(numParamBytesMangle >= sizeof(LPVOID));
+            numParamBytesMangle -= (UINT16)sizeof(LPVOID);
         }
-#endif // _TARGET_X86_
+
+        szProbedEntrypointName[-1] = '_';
```

Comment:
Logic is sound, but usage requires propagation of pointer arithmetic knowledge. Instead update the pointer to be correct and just use it naturally (i.e. `szProbedEntrypointName = szProbedEntrypointName - 1;`)